Add subsec to `ActiveSupport::TimeWithZone#inspect`

akinomaeni opened the pull request:
### Summary

<!-- Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together. -->

Add subsec to `ActiveSupport::TimeWithZone#inspect`.

before
```ruby
Time.at(1498099140).in_time_zone.inspect
# => "Thu, 22 Jun 2017 02:39:00 UTC +00:00"
Time.at(1498099140, 123456780, :nsec).in_time_zone.inspect
# => "Thu, 22 Jun 2017 02:39:00 UTC +00:00"
Time.at(1498099140 + Rational("1/3")).in_time_zone.inspect
# => "Thu, 22 Jun 2017 02:39:00 UTC +00:00"
```

after
```ruby
Time.at(1498099140).in_time_zone.inspect
# => "Thu, 22 Jun 2017 02:39:00.000000000 UTC +00:00"
Time.at(1498099140, 123456780, :nsec).in_time_zone.inspect
# => "Thu, 22 Jun 2017 02:39:00.123456780 UTC +00:00"
Time.at(1498099140 + Rational("1/3")).in_time_zone.inspect
# => "Thu, 22 Jun 2017 02:39:00.333333333 UTC +00:00"
```

### Other Information

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

#### Ruby 2.7 Time#inspect

It started inspecting sub-second of Time.
[Rdoc](https://ruby-doc.org/core-2.7.0/Time.html#inspect-method) [Ticket](https://bugs.ruby-lang.org/issues/15958) [Commit](https://github.com/ruby/ruby/commit/5208c431bef3240eb251f5da23723b324431a98e)

```shell
$ docker run ruby:2.7.0-alpine ruby -e 'p Time.at(1498099140)'
2017-06-22 02:39:00 +0000
$ docker run ruby:2.7.0-alpine ruby -e 'p Time.at(1498099140, 123456780, :nsec)'
2017-06-22 02:39:00.12345678 +0000
$ docker run ruby:2.7.0-alpine ruby -e 'p Time.at(1498099140 + Rational("1/3"))'
2017-06-22 02:39:00 1/3 +0000
```

```shell
$ docker run ruby:2.6.5-alpine ruby -e 'p Time.at(1498099140)'
2017-06-22 02:39:00 +0000
$ docker run ruby:2.6.5-alpine ruby -e 'p Time.at(1498099140, 123456780, :nsec)'
2017-06-22 02:39:00 +0000
$ docker run ruby:2.6.5-alpine ruby -e 'p Time.at(1498099140 + Rational("1/3"))'
2017-06-22 02:39:00 +0000
```

#### Motivation
`ActiveSupport::TimeWithZone#inspect` does not display sub-second. And it confuses me that times having different sub-seconds look like the same time.


- Database adapter: mysql2
- `create_table :songs { |t| t.datetime :released_at, precision: 6 }`

```ruby
# git branch => master
vagrant@rails-dev-box:/vagrant/my-test-app$ bin/rails c
Loading development environment (Rails 6.1.0.alpha)
irb(main):001:0> now = Time.current
=> Sat, 04 Apr 2020 13:37:58 UTC +00:00
irb(main):002:0> released_at = Song.new(released_at: now).released_at
=> Sat, 04 Apr 2020 13:37:58 UTC +00:00
irb(main):003:0> now == released_at
=> false
irb(main):004:0> now.nsec
=> 433841399
irb(main):005:0> released_at.nsec
=> 433841000
```

After the change it's clear.

```ruby
# git branch => inspect-time-with-zone-subsec
vagrant@rails-dev-box:/vagrant/my-test-app$ bin/rails c
Loading development environment (Rails 6.1.0.alpha)
irb(main):001:0> now = Time.current
=> Sat, 04 Apr 2020 13:39:12.231596578 UTC +00:00
irb(main):002:0> released_at = Song.new(released_at: now).released_at
=> Sat, 04 Apr 2020 13:39:12.231596 UTC +00:00
irb(main):003:0> now == released_at
=> false
```

akinomaeni committed:
Add subsec to `ActiveSupport::TimeWithZone#inspect`

Time.at(1498099140).in_time_zone.inspect
=> "Thu, 22 Jun 2017 02:39:00 UTC +00:00"
Time.at(1498099140, 123456780, :nsec).in_time_zone.inspect
=> "Thu, 22 Jun 2017 02:39:00.12345678 UTC +00:00"
Time.at(1498099140 + Rational("1/3")).in_time_zone.inspect
=> "Thu, 22 Jun 2017 02:39:00 1/3 UTC +00:00"

kaspth commented on activesupport/lib/active_support/time_with_zone.rb:
```diff
@@ -139,7 +139,7 @@ def zone
     #   Time.zone.now.inspect # => "Thu, 04 Dec 2014 11:00:25 EST -05:00"
     def inspect
-      "#{time.strftime('%a, %d %b %Y %H:%M:%S')} #{zone} #{formatted_offset}"
+      "#{time.strftime('%a, %d %b %Y %H:%M:%S')}#{formatted_subsec_with_delimiter} #{zone} #{formatted_offset}"
```
Can we get this via `strftime` in 2.7? Although it's potentially dangerous to make this Ruby version specific, there's a significant amount of code that it would be nice to trim away.

kaspth replied:
Looks like %L and %N are candidates:

```
  %L - Millisecond of the second (000..999)
  %N - Fractional seconds digits, default is 9 digits (nanosecond)
          %3N  millisecond (3 digits)   %15N femtosecond (15 digits)
          %6N  microsecond (6 digits)   %18N attosecond  (18 digits)
          %9N  nanosecond  (9 digits)   %21N zeptosecond (21 digits)
          %12N picosecond (12 digits)   %24N yoctosecond (24 digits)
```

akinomaeni replied:
@kaspth Thank you for your suggestion.

What do you think about changing the handling of rational subsecs that cannot be represented by integers even if subsec digits is increased?

---

Ruby implements its own subsec format without using strftime in Time#inspect.
https://github.com/ruby/ruby/blob/d827c718db32c07f025f88fd792e9407e19e66e1/time.c#L4143-L4157
So we cannot get the same formatted subsec via strftime.

#### case with the current `formatted_subsec_with_delimiter`. 
```
irb(main):001:0> t1 = Time.at(1498099140 + Rational("2/7")).in_time_zone
=> Thu, 22 Jun 2017 02:39:00 2/7 UTC +00:00
irb(main):002:0> t2 = Time.at(1498099140, 285714285, :nsec).in_time_zone
=> Thu, 22 Jun 2017 02:39:00.285714285 UTC +00:00
irb(main):003:0> t1 == t2
=> false
```


#### case with `%9N`
```
irb(main):001:0> t1 = Time.at(1498099140 + Rational("2/7")).in_time_zone
=> Thu, 22 Jun 2017 02:39:00.285714285 UTC +00:00
irb(main):002:0> t2 = Time.at(1498099140, 285714285, :nsec).in_time_zone
=> Thu, 22 Jun 2017 02:39:00.285714285 UTC +00:00
irb(main):003:0> t1 == t2
=> false
```

---

In my opinion, I want to know the contents exactly with inspect, different from to_s and other formatting methods.

But if Rails team thinks that maintainability is more important than accuracy here, I would support it.

akinomaeni replied:
Updated doc 899d006

akinomaeni replied:
@kaspth Here is the patch with `%9N` https://github.com/akinomaeni/rails/commit/7582ed74b6b7f26384194b8a773df812666df7bf. Should I push it to this PR?

kaspth replied:
Wow, that's cool! How does it look on Ruby 2.6? Feel free to push to here and get the build to show us.

akinomaeni replied:
```
irb(main):001:0> RUBY_VERSION
=> "1.9.3"
irb(main):002:0> Time.now.strftime('%9N')
=> "437936000"
```

`%9N` is already available on Ruby 1.9. So no worry on 2.6.

The new thing in Ruby 2.7 is this change.
> Ruby implements its own subsec format without using strftime in Time#inspect.
> https://github.com/ruby/ruby/blob/d827c718db32c07f025f88fd792e9407e19e66e1/time.c#L4143-L4157

akinomaeni replied:
> #### case with `%9N`
> ```
> irb(main):001:0> t1 = Time.at(1498099140 + Rational("2/7")).in_time_zone
> => Thu, 22 Jun 2017 02:39:00.285714285 UTC +00:00
> irb(main):002:0> t2 = Time.at(1498099140, 285714285, :nsec).in_time_zone
> => Thu, 22 Jun 2017 02:39:00.285714285 UTC +00:00
> irb(main):003:0> t1 == t2
> => false
> ```

This is on Ruby 2.5.5.

akinomaeni replied:
@kaspth Pushed!

kaspth replied:
> The new thing in Ruby 2.7 is this change.

Ah, I see what you mean. I think using `%9N` covers it just enough. The potential issue with `Rational` doesn't bother me with what you've shown me.

kaspth reviewed (commented):
Interesting idea! Hopefully this should make it more obvious when times are different in minitest assertions too.

akinomaeni committed:
Add subsec to `ActiveSupport::TimeWithZone#inspect`

Time.at(1498099140).in_time_zone.inspect
=> "Thu, 22 Jun 2017 02:39:00.000000000 UTC +00:00"
Time.at(1498099140, 123456780, :nsec).in_time_zone.inspect
=> "Thu, 22 Jun 2017 02:39:00.123456780 UTC +00:00"
Time.at(1498099140 + Rational("1/3")).in_time_zone.inspect
=> "Thu, 22 Jun 2017 02:39:00.333333333 UTC +00:00"

akinomaeni committed:
Show subsec of Time.zone.now [ci skip]

kaspth commented:
Thanks for working on this! 🙏
